### PR TITLE
Support unmount and remount in one atomic operation

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes are proposed in this pull request?

Support unmount and mount to another mount point within the same lock

### Why are the changes needed?

Support replace ufs for a mount point. The operation is atomic so no user operations should fail due to observing the middle state.

### Does this PR introduce any user facing changes?

A new rpc field and cmd options are added.

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 1]
```diff
@@ -3608,6 +3608,13 @@ public void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountContext conte
         }
         mMountTable.checkUnderWritableMountPoint(alluxioPath);
 
+        MountInfo mountInfo = mMountTable.getMountTable().get(alluxioPath);
```
Comment (jenoudet) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3611]:
```suggestion
        MountInfo mountInfo = mMountTable.getMountTable().get(alluxioPath.getPath());
```

**Comment by maobaolong** [comment]
@jenoudet Thanks for your review, PTAL.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 2]
```diff
@@ -3608,6 +3608,13 @@ public void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountContext conte
         }
         mMountTable.checkUnderWritableMountPoint(alluxioPath);
 
+        MountInfo mountInfo = mMountTable.getMountTable().get(alluxioPath.getPath());
+        Map<String, String> propertyMap = context.getOptions().getPropertiesMap();
+        if (mountInfo != null && propertyMap != null
+            && Boolean.parseBoolean(propertyMap.get("remount"))) {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3614]: per our discussion, let's change this to a new option in mount command

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the work! I added a comment.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 3]
```diff
@@ -3608,6 +3608,9 @@ public void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountContext conte
         }
         mMountTable.checkUnderWritableMountPoint(alluxioPath);
 
+        if (context.getOptions().getRemount()) {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3611]: check `hasRemount()` first, for backward compatibility?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 4]
```diff
@@ -3608,6 +3608,9 @@ public void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountContext conte
...
+          unmountInternal(rpcContext, inodePath);
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3612]: also add a log here so you can see

File: shell/src/main/java/alluxio/cli/fs/command/MountCommand.java [hunk 5]
```diff
@@ -51,6 +51,13 @@ public final class MountCommand extends AbstractFileSystemCommand {
           .hasArg(false)
           .desc("mount point is shared")
           .build();
+  private static final Option REMOUNT_OPTION =
+      Option.builder()
+          .longOpt("remount")
+          .required(false)
+          .hasArg(false)
+          .desc("unmount and mount together")
```
Comment (jiacheliu3) [inline shell/src/main/java/alluxio/cli/fs/command/MountCommand.java:59]:
```suggestion
          .desc("unmount and remount in one locked operation, with atomicity")
```

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
added some comments, thanks for the work!

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 3 above)
Comment (maobaolong) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3611]: image

This is a screenshot of generated class `MountPOptions`, it return the `readOnly_` directly.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 2 above)
Comment (maobaolong) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 4 above)
Comment (maobaolong) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:3614]: done

**Comment by maobaolong** [comment]
@jiacheliu3 Thank you for the review, PTAL

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**